### DISCUSSION TITLE
Type stable division

### DISCUSSION THREAD

**Discussion by vlulla** [duckdb/duckdb#15499] [Q&A] [answered]
Is it possible to have a type stable division operation? Specifically, how hard would it be to include an implementation of the `/` operator for `numeric` data type where the result type of the division operation is not a `double`? It appears that both [PostgreSQL](https://www.postgresql.org/docs/current/functions-math.html) and [BQ](https://cloud.google.com/bigquery/docs/reference/standard-sql/operators#arithmetic_operators) provide a type stable division operation for the arbitrary precision numeric data type.

The below queries demonstrate the type stability in these three databases:

```SQL
select typeof(cast(2.4 as numeric) / 1); -- duckdb
select `bigquery-public-data.persistent_udfs`.typeof(cast(2.4 as numeric) / 1); -- BQ
select pg_typeof(cast(2.4 as numeric) / 1); -- PostgreSQL
```

Thanks in advance!

**Comment by soerenwolfers** [comment]
In PG, 

```sql
SELECT (1.0::NUMERIC / 3.0::NUMERIC) * 3.0::NUMERIC;
```
```
0.999999999999999999990
```

Is it really helpful that the type of this is `NUMERIC`, which suggests more precision than is actually delivered?

You can always cast results of divisions back to decimals yourself if you want that. Quoting from the documentation:

> Division of fixed-point decimals does not typically produce numbers with finite decimal expansion

-- therefore, it doesn't make much sense to use the `NUMERIC` datatype for that, which is usually used for exact computations.

**Comment by vlulla** [comment]
@soerenwolfers fair point!

I use `NUMERIC(18,2)` to store monetary values. And, if i am not careful enough to ensure that monetary calculations contain only numeric types then there will be slight drift in calculations. Here's a small example to demonstrate what I mean:

```sql
-- Works for both PG an duckdb.
with 
    df0 as (select 1000*random() as money_col from generate_series(1,1000))
  , df1 as (select row_number() over() as rn,*,money_col::numeric(18,2) as money_col1 from df0)
  , df2 as (select *,sum(money_col) over w as cum_money_col,sum(money_col1) over w as cum_money_col1 from df1 window w as (ORDER BY rn ROWS BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)) 
select *,cum_money_col::numeric(18,2) as cum_money_col_cast from df2;
```

I merely proposed implementing this type stable division operation because BQ and PG offer this operation (regardless of whether this operation really has the necessary precision) and I was _surprised_ to find it missing in duckdb. This missing operation does not deter me from using duckdb so it's no hassle.

Anyways, thanks for your thoughts!

**Reply by soerenwolfers** [reply]
Does this example not preserve numeric type? If so, then I agree that it should.

**Reply by vlulla** [reply]
In this example it does. The subtlety is that if you change `sum` to `avg` (if you wanted to generate weekly/monthly/quarterly averages) then the numeric column will be automatically converted to double. Anyways, now that I'm aware of this subtlety i'll be careful.

**Reply by soerenwolfers** [reply, marked as answer]
Yeah, I maintain that that's an advantage: You are thereby warned that you're not actuallly getting exact results.
